feat: Add Documentation for Coder Modules (#164)
coder/registry: MODULE.md

                      │    Module Registry      │◄─── Publishes
                      │                         │     modules
                      │ Hosts and distributes   │
                      │    reusable modules     │
                      └───────────┬─────────────┘
                                  │
                                  │ Provides modules
                                  ▼
                      ┌─────────────────────────┐
                      │      Templates          │◄─── Users create
                      │                         │     templates
                      │ Define infrastructure   │
                      │ and integrate modules   │
                      └───────────┬─────────────┘
                                  │
                                  │ Provisions
                                  ▼
                      ┌─────────────────────────┐
                      │      Workspaces         │◄─── Developers use
                      │                         │     workspaces
                      │  Running development    │
                      │     environments        │
                      └─────────────────────────┘

> ```

Reviewer: Please replace with a rendered diagram. You can use tldraw or excalidraw for that.

### Terraform Foundation

> Coder modules are implemented as Terraform modules using HashiCorp's infrastructure-as-code framework. The [Coder Terraform provider](https://registry.terraform.io/providers/coder/coder/latest/docs) extends standard Terraform functionality with workspace-specific resources:

Reviewer: ```suggestion
Coder modules are written in Terraform and are used to extend Coder Templates. The [Coder Terraform provider](https://registry.terraform.io/providers/coder/coder/latest/docs) extends standard Terraform functionality with workspace-specific resources:
```
May be link to Coder templates

**Core Module Resources:**

- [`coder_script`](https://registry.terraform.io/providers/coder/coder/latest/docs/resources/script): Executes scripts during workspace lifecycle events
- [`coder_app`](https://registry.terraform.io/providers/coder/coder/latest/docs/resources/app): Creates accessible applications within the workspace interface
- [`coder_env`](https://registry.terraform.io/providers/coder/coder/latest/docs/resources/env): Sets environment variables in the workspace

**Template Resources (rarely used in modules):**

- [`coder_agent`](https://registry.terraform.io/providers/coder/coder/latest/docs/resources/agent): Defines agent configuration for workspace runtime
- [`coder_parameter`](https://registry.terraform.io/providers/coder/coder/latest/docs/resources/parameter): Collects user input during workspace provisioning
- [`coder_metadata`](https://registry.terraform.io/providers/coder/coder/latest/docs/resources/metadata): Displays resource information in the workspace interface

**Advanced Resources:**

> - [`coder_agent_instance`](https://registry.terraform.io/providers/coder/coder/latest/docs/resources/agent_instance): Manages agent instance connections

Reviewer: This is very rarely used so can be skipped.

- [`coder_ai_task`](https://registry.terraform.io/providers/coder/coder/latest/docs/resources/ai_task): Defines AI-powered task automation
- [`coder_devcontainer`](https://registry.terraform.io/providers/coder/coder/latest/docs/resources/devcontainer): Configures development container environments

Modules use standard Terraform syntax plus these Coder-specific resources to integrate with workspaces.

### Provider and Terraform Requirements

Modules specify required Terraform and provider versions:

```tf
terraform {
  required_version = ">= 1.0"

  required_providers {
    coder = {
      source  = "coder/coder"
      version = ">= 2.5"
    }
  }
}
```

This ensures compatibility with minimum versions of Terraform and the Coder provider.

**Current Version Standards:**

- **Latest Coder Provider**: v2.8.0 (June 2025)

> - **Recommended Minimum**: `>= 2.5` (used by most current modules)

Reviewer: ```suggestion
- **Recommended Minimum**: `>= 2.5` (used by most current modules)
```
There is no recommended version. It always depends on the feature you are using in your module and the minimum version of the provider where that feature is available.

- **Legacy Modules**: Some older modules still use lower versions (0.12+) but should be updated when possible

> New modules should use `>= 2.5` to ensure access to modern features like app groups, prebuilt workspaces, and improved validation.

Reviewer: ```suggestion
New modules should use `>= 2.5` to ensure access to modern features like app groups, prebuilt workspaces, and improved validation.
```
Again, we should not recommend 2.5; maybe we can recommend 2.0, as that is our current major release or remove this line in labor of explaining the minimum version.
